### PR TITLE
Line segment hashcode collision reduction

### PR DESCRIPTION
**Pull request by Skibisky**
### Description
Line Segment GetHashCode creates collisions.
Seems to be straight lines of same length that are 10 units away in 1 axis.

Added a very simple unit test.
Added a performance test which should identify if things go bad.

Could probably increase robustness by testing lines of varying length/offset/direction.

### CODE REVIEW

**Comment by DGuidi** [comment]
I would like to highlight two main issues with this pull request, that personally made me not comfortable this PR to be accepted

1. The current implementation is the same [JTS code uses](https://github.com/locationtech/jts/blob/149b3890780b42836cee64b0c99c1daecb085ba2/modules/core/src/main/java/org/locationtech/jts/geom/LineSegment.java#L643), so for personal feelings I'm against introduce code differences (but we can open an issue in JTS code, of course)
2. I'm not comfortable with _unchecked_ code: but I must admit that I'm not anymore aligned with .net intrinsics, so I'm not sure that unckecked code can introduce issues with different runtimes and/or target compilation: maybe unckecked is used in other parts of NTS codebase? If it's already used, of course this second point is not valid

**Comment by Skibisky** [comment]
Thanks for the link, I saw some references to JTS around. 
I will see if the hash collisions also occur in the Java environment.
The usage of 'unchecked' here permits integer overflow so that the runtime doesn't throw an exception.

**Comment by DGuidi** [comment]
What I mean is: using unchecked in a codebase allow an assembly to be used in different context, as example in xamarin forms targeting iOS?

**Comment by airbreather** [comment]
> What I mean is: using unchecked in a codebase allow an assembly to be used in different contexts, as example in xamarin forms targeting iOS?

`unchecked` should be fine.  It's actually the default, but this pattern is often used in places where overflow is **expected** to happen, since sometimes people turn on checking.

I would personally fix this by doing something more like
```csharp
return (_p0.X, _p0.Y, _p1.X, _p1.Y).GetHashCode();
```
but this works too

**Comment by Skibisky** [comment]
JTS have made changes to LineSegment hashCode, is the submission in an acceptable state to continue?

**Review by FObermaier — COMMENTED** [review]
Remove dead code.

File: src/NetTopologySuite/Geometries/LineSegment.cs [hunk 1]
```diff
@@ -700,6 +700,16 @@ public override string ToString()
...
+            {
+                int hash = 17;
+                hash = hash * 29 + _p0.X.GetHashCode( );
+                hash = hash * 29 + _p0.Y.GetHashCode( );
+                hash = hash * 29 + _p1.X.GetHashCode( );
+                hash = hash * 29 + _p1.Y.GetHashCode( );
+                return hash;
+            }
+
             long bits0 = BitConverter.DoubleToInt64Bits(_p0.X);
```
Comment (FObermaier) [inline src/NetTopologySuite/Geometries/LineSegment.cs:713]: Remove dead code

**Review by FObermaier — CHANGES_REQUESTED** [review]

**Review by FObermaier — APPROVED** [review]

**Comment by FObermaier** [comment]
Thanks for your contribution

**Pull request merged**